src/RecurringAllowanceManager.sol

```diff
@@ -0,0 +1,306 @@
+    /// @return currentCycle Currently active cycle with spend usage (struct).
+    function getCurrentCycle(RecurringAllowance calldata recurringAllowance) public view returns (CycleUsage memory) {
+        // check current timestamp is within recurring allowance time range
+        uint48 currentTimestamp = uint48(block.timestamp);
+        if (currentTimestamp < recurringAllowance.start) {
+            revert BeforeRecurringAllowanceStart(recurringAllowance.start);
+        } else if (currentTimestamp > recurringAllowance.end) {
+            revert AfterRecurringAllowanceEnd(recurringAllowance.end);
+        }
+
+        // return last cycle if still active, otherwise compute new active cycle start time with no spend
+        CycleUsage memory lastUpdatedCycle = _lastUpdatedCycle[getHash(recurringAllowance)][recurringAllowance.account];
+
+        // last cycle exists if spend is non-zero
+        bool lastCycleExists = lastUpdatedCycle.spend != 0;
+
+        // last cycle still active if current timestamp within [start, end - 1] range.
+        bool lastCycleStillActive =
+            currentTimestamp < uint256(lastUpdatedCycle.start) + uint256(recurringAllowance.period);
+
+        if (lastCycleExists && lastCycleStillActive) {
+            return lastUpdatedCycle;
+        } else {
+            // last active cycle does not exist or is outdated, determine current cycle
+
+            // current cycle progress is remainder of time since first recurring cycle mod reset period
+            uint48 currentCycleProgress = (currentTimestamp - recurringAllowance.start) % recurringAllowance.period;
+
+            // current cycle start is progress duration before current time
+            uint48 start = currentTimestamp - currentCycleProgress;
+
+            // current cycle end will overflow if period is sufficiently large
+            bool endOverflow = uint256(start) + uint256(recurringAllowance.period) > type(uint48).max;
+
+            // end is one period after start or maximum uint48 if overflow
+            uint48 end = endOverflow ? type(uint48).max : start + recurringAllowance.period;
+
+            return CycleUsage({start: start, end: end, spend: 0});
+        }
+    }
+
+    /// @notice Approve recurring allowance.
+    ///
+    /// @param recurringAllowance Details of the recurring allowance.
+    function _approve(RecurringAllowance calldata recurringAllowance) internal {
+        bytes32 hash = getHash(recurringAllowance);
+        _isApproved[hash][recurringAllowance.account] = true;
+        emit RecurringAllowanceApproved(hash, recurringAllowance.account, recurringAllowance);
+    }
+
+    /// @notice Withdraw tokens from account.
+    ///
+    /// @dev Function is virtual for easy overriding to support other account implementations.
+    ///
+    /// @param owner Account to withdraw tokens from.
+    /// @param recipient Account to withdraw tokens to.
+    /// @param token Address of token (either ether or ERC20 contract).
```

Comment:
This seems fine to me. So ERC-721 or 1155 will have to be follow ons?

Reply:
For now yes! I'm happy to build NFT support after we get some initial use cases and customers requesting it 👍